### PR TITLE
First pull request

### PR DESCRIPTION
**Pull request by Dumbldore**
GUI + opencv face recognition

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# Byte-compiled / optimized / DLL files
+__pycache__/
+*.py[cod]
+*$py.class
+
+# C extensions
+*.so
+
+# Distribution / packaging
```
Comment (pawelkopec) [inline .gitignore:6]: I guess we do not need so much stuff here, but I leave it to your choice.

File: app.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import argparse
+import numpy as np
+import cv2
+
+# creates a Flask application, named app
+app = Flask(__name__)
+
+
+cascpath = "haarcascade_frontalface_default.xml"
+modelpath = "model.hdf5"
```
Comment (pawelkopec) [inline app.py]: Why initialize here if you pass the path from cli?

File: app.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+graph = tf.get_default_graph()
+model = load_model('model.hdf5')
+model.load_weights('weights.03.hdf5')
+print('loaded everything :)')
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Input of model, weights and cascade path')
+    parser.add_argument('--cascade-path', '-c', required=False, type=str,
+                        help='OpenCV cascade file path, default set to: ' +
```
Comment (pawelkopec) [inline app.py]: No need to wrote what the default is. Argparser will write it. Additionally - why should we have a default param for that?

File: app.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                             cascpath, default=cascpath)
+    parser.add_argument('--model', '-t', required=False, type=str,
+                        help='Model address address, default' +
```
Comment (pawelkopec) [inline app.py]: Why "address"? I suggest "path" as above.

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1,33 @@
+absl-py==0.7.0
```
Comment (pawelkopec) [inline requirements.txt]: How did you get all those deps? Eg do we use click somewhere?

File: image_utils.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+def smile_detecting(face, model):
+    """ Detecting presence or lack of smile on the face and returning apopriate string """
+    face = face[:, :, 0]
+    mean_image = np.mean(face)
+    std_image = np.std(face)
+    face = (face - mean_image) / std_image
+    face = face.reshape(1, 48, 48, 1)
+    with graph.as_default():
+        prediction = model.predict(face)
+    if prediction[0, 0] >= 0.5:
```
Comment (pawelkopec) [inline image_utils.py]: Rather return a number that can be visualised in the front-end. Additionally lack of smile is not necessarily sadness.

File: requirements.txt (hunk 5 above)
Comment (Dumbldore) [reply requirements.txt]: We don't use click but it installs automatically with Flask

Comment (pawelkopec) [reply requirements.txt]: Why do we need it in requirements then?

File: image_utils.py (hunk 6 above)
Comment (Dumbldore) [reply image_utils.py]: done

File: app.py (hunk 4 above)
Comment (Dumbldore) [reply app.py]: done

File: app.py (hunk 3 above)
Comment (Dumbldore) [reply app.py]: done

File: app.py (hunk 2 above)
Comment (Dumbldore) [reply app.py]: done

File: .gitignore (hunk 1 above)
Comment (Dumbldore) [reply .gitignore:6]: done

File: requirements.txt (hunk 5 above)
Comment (Dumbldore) [reply requirements.txt]: I replaced it, done

File: image_utils.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
+import cv2
+import numpy as np
+import tensorflow as tf
+from typing import Tuple
+
+
+RGB_DIMENSION_COUNT = 3
+graph = tf.get_default_graph()
```
Comment (pawelkopec) [inline image_utils.py:7]: Why globally? You can move it to function.

File: image_utils.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            image if is_rgb(image) else to_rgb(image) for image in images
+        ]
+    # Convert to grayscale otherwise.
+    else:
+        new_images = [
+            image if is_gray(image) else to_gray(image)
+            for image in images
+        ]
+    return new_images
+
```
Comment (pawelkopec) [inline image_utils.py]: Too many newlines.

File: app.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import numpy as np
+import cv2
+
+# creates a Flask application, named app
+app = Flask(__name__)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Input of model, weights and cascade path')
+    parser.add_argument('-cascade', required=True, type=str,
```
Comment (pawelkopec) [inline app.py]: Nit - args longer than one letter usually begin with --, not -.

File: image_utils.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+graph = tf.get_default_graph()
+
+
+def smile_detecting(face, model):
+    """ Detecting presence or lack of smile on the face and returning apopriate string """
+    face = face[:, :, 0]
+    mean_image = np.mean(face)
+    std_image = np.std(face)
+    face = (face - mean_image) / std_image
+    face = face.reshape(1, 48, 48, 1)
```
Comment (pawelkopec) [inline image_utils.py]: Why 48 hardcoded?

File: image_utils.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+def to_gray(img: np.ndarray) -> np.ndarray:
+    """ Convert image to the grayscale color space. """
+    return cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
+
+
+def crop(img: np.ndarray, rect: Tuple[int, int, int, int]) -> np.ndarray:
+    """ Crop out input image's fragment specified by the rectangle. """
+    x, y, w, h = rect
+    return img[ x:x + w, y:y + h]
```
Comment (pawelkopec) [inline image_utils.py]: Space before x:x.

File: image_utils.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    )
+    faces_list = []
+    for (x, y, w, h) in faces:
+        faces_list.append(img[x:x + w, y:y + h])
+    return faces_list
+
+
+def face_as_net_input(face, image_size):
+    """ Converting given face to input valid to the model"""
+    input = convert_to_colorspace([face], "grayscale")[0]
```
Comment (pawelkopec) [inline image_utils.py]: Avoid "input", it is reserved in Python. You can use sth like face_input.

File: image_utils.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+def convert_to_colorspace(images, color_space="rgb"):
+    """
+    Convert list of input images to the target colorspace
+    :param images: list of images to convert
+    :param color_space: target colorspace
+    :return: converted images
+    """
+    # Convert to rgb space if requested.
+    if color_space == "rgb":
+        new_images = [
```
Comment (pawelkopec) [inline image_utils.py:81]: This is minor, but you can make it shorter:
```
if color_space == "rgb":
    return [...]
return [...]

File: templates/main.html [hunk 14]
```diff
@@ -0,0 +1,84 @@
+<html>
```
Comment (pawelkopec) [inline templates/main.html:5]: Please add todos for later commits:
1. Display rectangle with face detected.
2. Center the camera image.

File: image_utils.py (hunk 7 above)
Comment (Dumbldore) [reply image_utils.py:7]: done

File: image_utils.py (hunk 8 above)
Comment (Dumbldore) [reply image_utils.py]: done

File: app.py (hunk 9 above)
Comment (Dumbldore) [reply app.py]: done

File: image_utils.py (hunk 10 above)
Comment (Dumbldore) [reply image_utils.py]: done

File: image_utils.py (hunk 11 above)
Comment (Dumbldore) [reply image_utils.py]: done

File: image_utils.py (hunk 12 above)
Comment (Dumbldore) [reply image_utils.py]: done

File: image_utils.py (hunk 13 above)
Comment (Dumbldore) [reply image_utils.py:81]: done

File: templates/main.html (hunk 14 above)
Comment (Dumbldore) [reply templates/main.html:5]: done

File: image_utils.py (hunk 7 above)
Comment (pawelkopec) [reply image_utils.py:7]: Where?

File: image_utils.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+import tensorflow as tf
+from typing import Tuple
+
+
+graph = tf.get_default_graph()
+
+
+def smile_detecting(face, model):
+    """ Detecting presence or lack of smile on the face and returning apopriate string """
+    neural_network_input_size = 48
```
Comment (pawelkopec) [inline image_utils.py:12]: You can infer it from model, but I leave it to your choice.

**Review by pawelkopec — APPROVED** [review]

**Pull request merged**